File: 24.swap-nodes-in-pairs.c
/*
 * @lc app=leetcode id=24 lang=c
 *
 * [24] Swap Nodes in Pairs
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* swapPairs(struct ListNode* head){
    if(head == NULL)
        return head;
    if(head->next == NULL)
        return head;
    struct ListNode* dummy = (struct ListNode*)malloc(sizeof(struct ListNode));
    dummy->next = head;
    struct ListNode* pre = dummy;
    while(pre->next && pre->next->next){
        struct ListNode* cur = pre->next;
        struct ListNode* post = pre->next->next;
        pre->next = post;
        cur->next = post->next; 
        post->next = cur;
        pre = pre->next->next;
    }
    return dummy->next;
}

File: 24.swap-nodes-in-pairs.py
#
# @lc app=leetcode id=24 lang=python3
#
# [24] Swap Nodes in Pairs
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head:
            return head
        if not head.next:
            return head
        lastNode = ListNode(0)
        lastNode.next = head
        cur = head
        head = head.next
    
        while cur and cur.next:
            lastNode.next = cur.next
            temp = cur.next.next
            cur.next.next = cur
            cur.next = temp
            lastNode = cur
            cur = temp
        return head
            